File: frontend/types/counterparty.ts
export type Counterparty = {
  id: number;
  user_id: number;
  name: string;
  opening_receivable_amount: number;
  opening_payable_amount: number;
  receivable_amount: number;
  payable_amount: number;
  created_at: string;
  updated_at: string;
};

export type CreateCounterpartyPayload = {
  name: string;
  opening_balance?: number;
  opening_balance_kind?: 'receivable' | 'payable';
};

File: app/models/counterparty.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Numeric, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class Counterparty(Base):
    __tablename__ = "counterparties"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    opening_receivable_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False, default=0, server_default="0")
    opening_payable_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False, default=0, server_default="0")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    user = relationship("User", back_populates="counterparties")
    transactions = relationship("Transaction", back_populates="counterparty")
